crates/ide_assists/src/handlers/generate_constant.rs

```diff
@@ -0,0 +1,130 @@
+use crate::assist_context::{AssistContext, Assists};
+use hir::HirDisplay;
+use ide_db::{
+    assists::{AssistId, AssistKind},
+    defs::NameRefClass,
+};
+use syntax::{
+    ast::{self, edit::IndentLevel},
+    AstNode,
+};
+
+// Assist: generate_constant
+//
+// Generate a named constant.
+//
+// ```
+// struct S { i: usize }
+// impl S { pub fn new(n: usize) {} }
+// fn main() {
+//     let v = S::new(CAPA$0CITY);
+// }
+// ```
+// ->
+// ```
+// struct S { i: usize }
+// impl S { pub fn new(n: usize) {} }
+// fn main() {
+//     const CAPACITY: usize = $0;
+//     let v = S::new(CAPACITY);
+// }
+// ```
+
+pub(crate) fn generate_constant(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let statement = ctx.find_node_at_offset::<ast::Stmt>()?;
+    let arg_list = ctx.find_node_at_offset::<ast::ArgList>()?;
+    let expr = ctx.find_node_at_offset::<ast::Expr>()?;
+    let constant_token = ctx.find_node_at_offset::<ast::NameRef>()?;
```

Comment:
```suggestion
    let constant_token = ctx.find_node_at_offset::<ast::NameRef>()?;
    let expr = constant_token.syntax().ancestors().find_map(ast::Expr::cast)?;
    let statement = expr.syntax().ancestors().find_map(ast::Stmt::cast)?;
```
Better to use `find_node_at_offset` only once to prevent us from grabbing unrelated nodes.